File: src/test/java/test/FileTest.java
package test;

import com.github.t1.yaml.Yaml;
import com.github.t1.yaml.model.Document;
import com.github.t1.yaml.model.ScalarNode;
import com.github.t1.yaml.model.Stream;
import lombok.SneakyThrows;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;

import static java.nio.charset.StandardCharsets.UTF_8;
import static org.assertj.core.api.Assertions.assertThat;

class FileTest {
    private Path yamlFile;

    private static Document expectedDocument;
    private static Stream expectedStream;

    private static Document document;
    private static Stream stream;

    @BeforeEach @SneakyThrows(IOException.class) void setup() {
        yamlFile = Files.createTempFile("test-", ".yaml");
        Files.write(yamlFile, "dummy-string".getBytes(UTF_8));

        expectedDocument = new Document().node(new ScalarNode().line("dummy-string"));
        expectedStream = new Stream().document(expectedDocument);
    }


    interface ThenShouldMatchDocument {
        @Test default void thenShouldMatchDocument() { assertThat(document).isEqualTo(expectedDocument); }
    }

    interface ThenShouldMatchStream {
        @Test default void thenShouldMatchStream() { assertThat(stream).isEqualTo(expectedStream); }
    }


    @Nested class givenReader {
        BufferedReader reader;

        @BeforeEach @SneakyThrows(IOException.class) void setUp() { reader = Files.newBufferedReader(yamlFile, UTF_8); }

        @Nested class whenParseFirst implements ThenShouldMatchDocument {
            @BeforeEach void setup() { document = Yaml.parseFirst(reader); }
        }

        @Nested class whenParseSingle implements ThenShouldMatchDocument {
            @BeforeEach void setup() { document = Yaml.parseSingle(reader); }
        }

        @Nested class whenParseAll implements ThenShouldMatchStream {
            @BeforeEach void setup() { stream = Yaml.parseAll(reader); }
        }
    }


    @Nested class givenInputStream {
        InputStream stream;

        @BeforeEach @SneakyThrows(IOException.class) void setUp() { stream = Files.newInputStream(yamlFile); }

        @Nested class whenParseFirst implements ThenShouldMatchDocument {
            @BeforeEach void setup() { document = Yaml.parseFirst(stream); }
        }

        @Nested class whenParseSingle implements ThenShouldMatchDocument {
            @BeforeEach void setup() { document = Yaml.parseSingle(stream); }
        }

        @Nested class whenParseAll implements ThenShouldMatchStream {
            @BeforeEach void setup() { FileTest.stream = Yaml.parseAll(stream); }
        }
    }
}

File: src/test/java/test/FileTest.kt
package test

import com.github.t1.yaml.Yaml
import com.github.t1.yaml.model.Document
import com.github.t1.yaml.model.ScalarNode
import com.github.t1.yaml.model.Stream
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import java.io.BufferedReader
import java.io.InputStream
import java.nio.charset.StandardCharsets.UTF_8
import java.nio.file.Files
import java.nio.file.Path

@Suppress("ClassName")
class FileTest {
    private lateinit var yamlFile: Path

    @BeforeEach fun setup() {
        yamlFile = Files.createTempFile("test-", ".yaml")!!
        Files.write(yamlFile, "dummy-string".toByteArray(UTF_8))

        expectedDocument = Document().node(ScalarNode().line("dummy-string"))
        expectedStream = Stream().document(expectedDocument)
    }


    interface ThenShouldMatchDocument {
        @Test fun thenShouldMatchDocument() {
            assertThat(document).isEqualTo(expectedDocument)
        }
    }

    interface ThenShouldMatchStream {
        @Test fun thenShouldMatchStream() {
            assertThat(stream).isEqualTo(expectedStream)
        }
    }


    @Nested inner class givenReader {
        lateinit var reader: BufferedReader

        @BeforeEach fun setup() {
            reader = Files.newBufferedReader(yamlFile, UTF_8)!!
        }

        @Nested inner class whenParseFirst : ThenShouldMatchDocument {
            @BeforeEach fun setup() {
                document = Yaml.parseFirst(reader)
            }
        }

        @Nested inner class whenParseSingle : ThenShouldMatchDocument {
            @BeforeEach fun setup() {
                document = Yaml.parseSingle(reader)
            }
        }

        @Nested inner class whenParseAll : ThenShouldMatchStream {
            @BeforeEach fun setup() {
                stream = Yaml.parseAll(reader)
            }
        }
    }


    @Nested inner class givenInputStream {
        lateinit var stream: InputStream

        @BeforeEach fun setup() {
            stream = Files.newInputStream(yamlFile)!!
        }

        @Nested inner class whenParseFirst : ThenShouldMatchDocument {
            @BeforeEach fun setup() {
                document = Yaml.parseFirst(stream)
            }
        }

        @Nested inner class whenParseSingle : ThenShouldMatchDocument {
            @BeforeEach fun setup() {
                document = Yaml.parseSingle(stream)
            }
        }

        @Nested inner class whenParseAll : ThenShouldMatchStream {
            @BeforeEach fun setup() {
                FileTest.stream = Yaml.parseAll(stream)
            }
        }
    }

    companion object {

        private var expectedDocument: Document? = null
        private var expectedStream: Stream? = null

        private var document: Document? = null
        private var stream: Stream? = null
    }
}
